Cookbook/Proposals/Base class for FeedbackLoops systems.md

```diff
@@ -0,0 +1,157 @@
+  private let input = Feedback<State, Update>.input()
+  private let actionTransformer: (Action) -> Event
+
+  public init(
+    initial: State,
+    reducer: @escaping (State, Event) -> State,
+    feedbacks: [Feedback<State, Event>],
+    scheduler: DateScheduler,
+    actionTransformer: @escaping (Action) -> Event
+  ) {
+    self.actionTransformer = actionTransformer
+    self.state = Property(
+      initial: initial,
+      scheduler: scheduler,
+      reduce: { (state: State, update: Update) -> State in
+        switch update {
+        case let .event(event):
+          return reducer(state, event)
+        case let .mutation(mutation):
+          return mutation.mutate(state)
+        }
+      },
+      feedbacks: feedbacks.map { $0.mapEvent(Update.event) }
+        .appending(self.input.feedback)
+    )
+  }
+
+  open func send(action: Action) {
+    self.input.observer(.event(actionTransformer(action)))
+  }
+
+  open func mutate<V>(keyPath: WritableKeyPath<State, V>, value: V) {
+    self.input.observer(.mutation(Mutation(keyPath: keyPath, value: value)))
+  }
+
+  private enum Update {
+    case event(Event)
+    case mutation(Mutation)
+  }
+
+  private struct Mutation {
+    let mutate: (State) -> State
+
+    init<V>(keyPath: WritableKeyPath<State, V>, value: V) {
+      self.mutate = { state in
+        var copy = state
+
+        copy[keyPath: keyPath] = value
+
+        return copy
+      }
+    }
+  }
+}
+```
+
+2. Have one entity that both contains the State and can dispatch the action e.g [StateSnapshot](https://github.com/andersio/Cycler/blob/master/View/BoundView.swift#L34)
+```swift
+//@dynamicMemberLookup
+public struct StateSnapshot<T: FeedbackLoopType> {
+  public let state: T.State
+  private weak var feedbackLoop: T?
+
+  public init(state: T.State, feedbackLoop: T) {
+    self.state = state
+    self.feedbackLoop = feedbackLoop
+  }
+
+  public subscript<U>(dynamicMember keyPath: KeyPath<T.State, U>) -> U {
+    get {
+      return state[keyPath: keyPath]
+    }
+  }
+
+  public func send(action: T.Action) {
+    feedbackLoop?.send(action: action)
+  }
+
+  public func mutate<V>(keyPath: WritableKeyPath<T.State, V>, value: V) {
+    feedbackLoop?.mutate(keyPath: keyPath, value: value)
+  }
+}
+```
+Then our render would look like:
+
+```swift
+public protocol Renderer {
+  associatedtype FeedbackLoop: FeedbackLoopType
+  associatedtype Appearance: BoxAppearance
+  associatedtype Config
+
+  init(appearance: Appearance, config: Config)
+
+  func render(snapshot: StateSnapshot<FeedbackLoop>) -> AnyRenderable
+}
+```
+
+This will allow us easier bind VM to a component e.g
+
+```swift
+return Component(textDidChange: {
+   //This will still couse rerendering but we would not need to create an Event for it
+    snapshot.mutate(\.userName, value: $0)
+})
+```
+
+
+## Impact on the existing codebase
+
+This is additive no impact can be bundled together with new `RenderableViewController`
```

Comment:
Can we rephrase this? Sounds confusing imho

Reply:
We have a `RenderableViewController` which is suppose to replace `BoxViewController`